File: teachprogramming/static/projects/mini/digit_word.py

DIGIT_WORDS = {
    1: 'ONE',
    2: 'TWO',
    3: 'THREE',
    4: 'FOUR',
    5: 'FIVE',
    6: 'SIX',
    7: 'SEVEN',
    8: 'EIGHT',
    9: 'NINE',
}
def digit_word(s):
    """
    >>> digit_word('BOUNCE')
    1
    >>> digit_word('ENCODE')
    'NO'
    >>> digit_word('EIGHT')
    8
    >>> digit_word('BLACKJACK')
    'NO'
    >>> digit_word('FABULOUS')
    'NO'
    >>> digit_word('EXERCISE')
    'NO'
    >>> digit_word('DRIFTWOOD')
    2
    >>> digit_word('SERVICEMAN')
    7
    >>> digit_word('INSIGNIFICANCE')
    9
    >>> digit_word('THROWDOWN')
    2
    """
    assert s.isupper()
    for digit_int, digit_str in DIGIT_WORDS.items():
        digit_str = list(digit_str)
        for letter in s:
            if letter == digit_str[0]:
                digit_str.pop(0)
            if not digit_str:
                return digit_int
    return 'NO'

File: teachprogramming/static/projects/british_infomatics_olimpyad/DigitWord.java
package britishinfomaticsolymiad;

import java.util.ArrayList;
import java.util.List;

public class DigitWord {
  private static final List<String> lookup = new ArrayList<String>();

  private static void init() {
    lookup.add("ZERO");
    lookup.add("ONE");
    lookup.add("TWO");
    lookup.add("THREE");
    lookup.add("FOUR");
    lookup.add("FIVE");
    lookup.add("SIX");
    lookup.add("SEVEN");
    lookup.add("EIGHT");
    lookup.add("NINE");
  }

  public static String findDidget(String word) {
    System.out.println("Finding: "+word);
    //System.out.println("Size: "+lookup.size());
    for (int digit_number_index=0 ; digit_number_index<lookup.size() ; digit_number_index++) {
      String digit_string = lookup.get(digit_number_index);
      //System.out.println("Finding word: "+digit_string);
      int lookup_pointer = 0;
      for (int i=0 ; i<word.length() ; i++) {
        //System.out.println("Comparing "+word.charAt(i) +" with "+digit_string.charAt(lookup_pointer));
        if (word.charAt(i)==digit_string.charAt(lookup_pointer)) {
          lookup_pointer++;
          //System.out.println("Next!");
          if (lookup_pointer>=digit_string.length()) {
            return ""+digit_number_index;
          }
        }
      }
    }
    return "NO";
  }

  public static void main(String[] args) {
    init();
    System.out.println(DigitWord.findDidget("BOUNCE"));
    System.out.println(DigitWord.findDidget("ENCODE"));
    System.out.println(DigitWord.findDidget("EIGHT"));
    System.out.println(DigitWord.findDidget("BLACKJACK"));
    System.out.println(DigitWord.findDidget("FABULIOUS"));
    System.out.println(DigitWord.findDidget("EXERCISE"));
    System.out.println(DigitWord.findDidget("DRIFTWOOD"));
    System.out.println(DigitWord.findDidget("SERVICEMAN"));
    System.out.println(DigitWord.findDidget("INSIGNIFICANCE"));
    System.out.println(DigitWord.findDidget("THROWDOWN"));
  }

}
